wallet/src/main/java/com/android/identity_credential/wallet/ui/destination/provisioncredential/EvidenceRequest.kt

```diff
@@ -660,8 +839,88 @@ fun EvidenceRequestIcaoNfcTunnelView(
+        if (!showSuccessfulScanningScreen) {
+            EvidenceRequestIcaoView(tunnelScanner, permissionTracker, initialRoute) {
+                // upon completion of nfc scanning, show a transitional "Success" screen
+                showSuccessfulScanningScreen = true
+            }
+        } else { // user successfully scanned an NFC document
+            // briefly (1 sec) show a transitional success screen then finish the tunneling process
+            DocumentSuccessfullyScannedScreen(pauseDurationMs = 1000L) {
+                provisioningViewModel.finishTunnel()
+            }
+        }
+    }
+}
+
+/**
+ * Briefly show the "Document Scanning Successful" screen before moving to the next screen/step
+ * of evidence request (automatically after waiting for [pauseDurationMs]).
+ */
+@Composable
+fun DocumentSuccessfullyScannedScreen(pauseDurationMs: Long, onFinishedShowingScreen: () -> Unit) {
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(horizontal = 10.dp)
+    ) {
+        Image(
+            painter = painterResource(id = R.drawable.id_card),
+            contentDescription = stringResource(
+                R.string.accessibility_artwork_for,
+                stringResource(id = R.string.evidence_nfc_scan_successful_title)
+            ),
+            modifier = Modifier
+                .size(48.dp)
+                .padding(start = 16.dp)
+        )
+        Row(
+            modifier = Modifier.fillMaxWidth(),
+            horizontalArrangement = Arrangement.Center
+        ) {
+            Text(
+                text = stringResource(R.string.evidence_nfc_scan_successful_title),
+                modifier = Modifier.padding(16.dp),
+                textAlign = TextAlign.Left,
+                fontWeight = FontWeight.Bold,
+                style = MaterialTheme.typography.headlineLarge
+            )
+        }
+        Row(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(top = 100.dp, bottom = 200.dp),
+            horizontalArrangement = Arrangement.Center
+        ) {
+            Column(
+                modifier = Modifier.padding(top = 80.dp, bottom = 50.dp),
+                horizontalAlignment = Alignment.CenterHorizontally
+            ) {
+                Icon(
+                    painter = painterResource(id = R.drawable.check_circled),
+                    tint = colorResource(id = R.color.success_green),
+                    modifier = Modifier.size(100.dp),
+                    contentDescription = stringResource(
+                        R.string.accessibility_artwork_for,
+                        stringResource(id = R.string.evidence_nfc_scan_successful_title)
+                    )
+                )
+            }
+        }
+    }
+    LaunchedEffect(key1 = true) {
+        delay(pauseDurationMs)
+        onFinishedShowingScreen.invoke()
```

Comment:
No need for .invoke.
onFinishedShowingScreen()

Reply:
Technically the correct approach would be to use `.invoke()` on lambdas (such as callbacks) 
- https://kotlinlang.org/docs/lambdas.html#invoking-a-function-type-instance

Under the hood [Kotlin calls `invoke` on all function calls](https://kotlinlang.org/docs/operator-overloading.html#invoke-operator)
![Screenshot 2024-11-06 at 6 05 13 PM](https://github.com/user-attachments/assets/a034b346-c1c9-479b-bf31-714f9aa091e9)

Reply:
I see all that, and I still don't understand why we'd call invoke.

If Kotlin is calling invoke for us automatically, why are we making our code (marginally) more complex to read than it needs to be? The documentation you linked to starts with: "A value of a function type can be invoked by using its invoke(...) operator: f.invoke(x) or just f(x)."

That last "...or just f(x)" seems like the simplest way to call any function. Why would we ever explicitly add the invoke call that the compiler will add for us automatically?

If we're calling invoke instead, why not do it everywhere?
    LaunchedEffect.invoke(key1 = true) {
        delay.invoke(pauseDurationMs)
        onFinishedShowingScreen.invoke()

That just makes the code ugly, and it's unnecessary. It works, but it doesn't add anything. Why would we ever call invoke()?

Reply:
It's a tradeoff between readability and Kotlin DSL idiom. I'd rather be correct if it doesn't add too much of a burden to the reader.

My rule of thumb is to use `invoke` when dealing with lambdas -- especially if the lambda could be `null`. This immediately tells me that the function being called is a lambda object and not strongly defined. Ex:

```
fun foobar(lazyCallBack: (()->Unit)?=null){
  lazyCallBack?.invoke() 
}
```

And for normal functions I like to make direct function calls because they are more readable and calling them with `invoke` is superfluous. Ex

```
foobar(..)
```
or better yet with the example above it makes even more sense to make a direct function call

```
foobar {
   // do work when callback is called
}

```

Yes this is a matter of preference however I've read a lot of Kotlin and Android source code and have seen calls to `invoke()` when dealing with lambdas. If Kotlin/Android core functionality is defining a style in (core) components that are used by everyone who programs in Kotlin/Android, it is reasonable to also follow this practice in your app-wide code. 

Take a look at a sample of `NavHost` as the fastest example I could think of:

![Screenshot 2024-11-08 at 4 04 44 AM](https://github.com/user-attachments/assets/64523f3a-d8f1-4bbe-8f50-3197b64ee2d3)

Reply:
Ah, an idiom..  I'm all for following established styles and idioms.

And thanks for the elvis operator example when it's possibly null. That's the first case I can see where it actually makes complete sense to use the invoke function.

However... You keep using the term lambda here. But whether an object is a function or a lambda is based on where that object is defined, not where it's called. Looking at our onFinishedShowingScreen object here, all we know is it matches the type: () -> Unit. Theoretically, there are at least two ways we could define that:

As a regular function, which we then pass to DocumentSuccessfullyScannedScreen():
fun onFinished() { /* do stuff */ }
DocumentFullyScannedScreen(1000, onFinished)

(I recall you did that in a recent PR when a local callback function moved to a top-level function)

As a lambda, at the time we call DocumentSuccessfullyScannedScreen:
DocumentSuccessfullyScannedScreen(1000) {
  /* do stuff */
}

or maybe:

DocumentSuccessfullyScannedScreen(1000, onFinishedShowingScreen = { something something...})

By convention, we'll pretty much always do the second one, which makes it a lambda... But at the site where onFinishedShowingScreen is called, we really have no way to know whether it's a function, a lambda, or something else (eg. an instance of a class that's defined the invoke() operator). Nor should we care, because we can call all of those just like any function. And it's a little less predictable when there are multiple function parameters: there's more of a chance that one of the callbacks could be something other than a lambda.

So it still doesn't make any sense to me why we'd call invoke() in any case except the nullable one, with an elvis operator.

So we're back to the idiom rationale. You have a much more established history with Kotlin than I do, so I can accept that. That being said, I can't find any documentation or online discussions that indicate this is a style we should be following. Do you know of any?

And do you have a link to the source code from that screenshot? I'm having a hard time finding it.

Reply:
Good points! I think the set of conditions that comprise my rule of thumb for `invoke()` is whether the function being called is passed-in as a function type parameter (dynamic calls). This is a Kotlin functional programming paradigm that I've seen in a few Android/Kotlin core sources.

For example, in your sample code above you're attempting to pass the `onFinished` function when you call `DocumentFullyScannedScreen()` but the function `onFinished` is not defined as a `val` and you cannot pass in the function by name like that (which could infer that you're attempting to execute the function `onFinished` at the time of calling the composable function `DocumentFullyScannedScreen`, rather than pass it by reference so it can be called from inside the composable)
```
fun onFinished() { /* do stuff */ }
DocumentFullyScannedScreen(1000, onFinished) <--- doesn't work
```

Passing the function `onFinished` by reference:
```
DocumentFullyScannedScreen(1000, ::onFinished)
```

At this point, inside the scope of `DocumentFullyScannedScreen`, the function signature name `fun onFinished()` is irrelevant since the function is being passed by reference and behaves like a lambda expression because the signature of composable function `DocumentFullyScannedScreen` is accepting any function type parameter `()-> Unit` for the 2nd parameter (that is tied to whatever locally declared parameter name, such as `doWorkWhenFinished : ()->Unit`). 

So when calling the functional parameter `doWorkWhenFinished` it doesn't matter whether the passed-in value is a lambda expression or a function `onFinished` that is defined as a `val` or as `fun` -- a more Kotlin DSL idiomatic approach would be to call `doWorkWhenFinished.invoke()`. 

However, if we consider the function `onFinished` to be declared as a function type object
```
val onFinished : ()-> Unit = { ... }
```
And if it is being called directly (rather than passing it around to call dynamically), I would be more inclined to use `onFinished()`. Even though it is a function type object (without a function signature), it is not being called dynamically by passing a reference of it into some function. The same applies when `onFinished` has been declared with a function signature such `fun onFinished()`.

This automatically infers to the reader that the function `onFinished` has been locally declared somewhere (as a function-type object or function) and is not a lambda expression nor is it dynamically called.

The sample code I pasted above can be found by going to the `NavHost` file in AndroidStudio
![Screenshot 2024-11-12 at 10 04 40 AM](https://github.com/user-attachments/assets/cf51f592-86f0-439b-9aaa-09fe18757765)

Here's the snippet with the line numbers
![Screenshot 2024-11-12 at 10 44 38 AM](https://github.com/user-attachments/assets/3d07feda-857d-4f91-8c50-917c77ee2f85)

and the `NavHost` function parameter types such as `popExitTransition` are **not** defined to accept `null` values
![Screenshot 2024-11-12 at 11 33 56 AM](https://github.com/user-attachments/assets/3f52402e-3f0d-4004-9097-2dc34e3fc02f)


Here's another example, `TextView.kt` belonging to `package androidx.core.widget` that makes a distinction between calling referenced functions via `invoke()` (green borders) and calling a function directly (orange border)
![Screenshot 2024-11-12 at 11 59 57 AM](https://github.com/user-attachments/assets/90bad59c-c41e-45c0-a9f0-642f90ce6fca)

Having said that **however**, I've also seen core source code that directly calls referenced function type objects such as all the [scope functions](https://kotlinlang.org/docs/scope-functions.html); take `.also { .. }` and the referenced function object `block: (T) -> Unit`

![Screenshot 2024-11-12 at 12 51 53 PM](https://github.com/user-attachments/assets/928883c7-eef6-4897-b916-b21ff358998c)

I haven't been able to find any explicit documentation outlining the functional programming paradigms around this topic. 

From what I've seen (in Android/Kotlin core sources), and my rule of thumb for providing a clear way to distinguish between referenced functions or function-type objects and non-referenced functions/function-type objects is by calling `.invoke()` on dynamic calls of referenced functions or function-type objects.